File: go/ink/attestation.go
package ink

import (
	"crypto/ed25519"
	"encoding/base64"
	"regexp"
)

// The evidence primitive of specs/ink-attestation.md: a signed claim by one
// principal about another, verified from raw bytes, judged only by receiver
// policy. Base verification is signature, shape and window; there is no
// audience, no replay and no judgment of issuer or claim.

var attestationTopLevelKeys = map[string]bool{
	"protocol": true, "type": true, "issuer": true, "subject": true,
	"claimType": true, "claim": true, "attestationId": true,
	"issuedAt": true, "expiresAt": true, "signature": true,
}

var attestationRequiredKeys = []string{
	"protocol", "type", "issuer", "subject", "claimType", "claim",
	"attestationId", "issuedAt", "expiresAt", "signature",
}

const (
	attestationIDMin  = 16
	attestationIDMax  = 256
	claimTypeMin      = 3
	claimTypeMax      = 128
	attestationIDMaxs = 512 // issuer and subject share the DID/agent-id bound
)

// MaxAttestationBodyBytes is the byte ceiling on a raw attestation body,
// enforced before decoding, matching the reference MAX_ATTESTATION_BODY_BYTES.
const MaxAttestationBodyBytes = 65536

var claimTypeRe = regexp.MustCompile(`^[a-z0-9]+(\.[a-z0-9_]+)+$`)
var attestationIDRe = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

// AttestationReason is the stable rejection discriminator, matching the
// reference verifier's reason strings.
type AttestationReason string

const (
	AttestationReasonNone        AttestationReason = ""
	AttestationReasonSchema      AttestationReason = "schema"
	AttestationReasonSignature   AttestationReason = "signature"
	AttestationReasonNotYetValid AttestationReason = "not_yet_valid"
	AttestationReasonExpired     AttestationReason = "expired"
)

// VerifyAttestation verifies an attestation from its raw bytes against the
// resolved issuer public key and the verifier clock. It mirrors the TypeScript
// verifyAttestation byte for byte: byte cap, the shared signed-body parse,
// structural bounds, strict schema, a body signature over "tulpa/sign\n" +
// JCS(attestation without signature) under RFC 8032 strict Ed25519, then the
// validity window with an inclusive lower and exclusive upper bound. It fails
// closed with a typed reason on the first failure.
func VerifyAttestation(raw []byte, issuerPublicKey []byte, now string) (bool, AttestationReason) {
	reason, _ := verifyAttestationWith(raw, fixedKey(issuerPublicKey), now)
	return reason == AttestationReasonNone, reason
}

// AttestationKeyResult is the result of VerifyAttestationWithKeys: whether
// the attestation verified, the typed rejection reason on failure, and the
// candidate key that verified it on success.
type AttestationKeyResult struct {
	OK     bool
	Reason AttestationReason
	Key    MultiKeyResult
}

// VerifyAttestationWithKeys verifies an attestation from its raw bytes
// against a rotation-aware candidate issuer key set. Security
// considerations §"Issuer key rotation" (ink-attestation.md): an
// attestation verifies under the same rotation rules as any other signed
// body. A retired issuer key still verifies an attestation issued inside
// that key's validity window, and a revoked issuer key never verifies,
// even for an attestation whose issuedAt predates the revocation.
//
// The artifact clock is issuedAt, the moment the issuer signed the claim,
// not now (the verifier's clock, used only to judge freshness against
// expiresAt). Same code path as VerifyAttestation; only the signature step
// is rotation-aware.
func VerifyAttestationWithKeys(raw []byte, keys []CandidateKey, now string, hintKeyID string) AttestationKeyResult {
	reason, key := verifyAttestationWith(raw, candidateKeys(keys, hintKeyID), now)
	if reason != AttestationReasonNone {
		return AttestationKeyResult{Reason: reason}
	}
	return AttestationKeyResult{OK: true, Reason: reason, Key: key}
}

func verifyAttestationWith(raw []byte, s signerStrategy, now string) (AttestationReason, MultiKeyResult) {
	if len(raw) > MaxAttestationBodyBytes {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	obj, okParse := ParseSignedObject(raw)
	if !okParse {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	if !withinBodyBounds(obj) {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	signature, ok := validateAttestation(obj)
	if !ok {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	if !signatureRe.MatchString(signature) {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	if !s.keyOK() {
		return AttestationReasonSignature, MultiKeyResult{}
	}
	var artifactMs int64
	if s.needsClock {
		issuedAt, _ := obj["issuedAt"].(string)
		ms, ok := ParseInkTimestampMs(issuedAt)
		if !ok {
			return AttestationReasonSchema, MultiKeyResult{}
		}
		artifactMs = ms
	}
	unsigned := make(map[string]interface{}, len(obj))
	for k, v := range obj {
		if k != "signature" {
			unsigned[k] = v
		}
	}
	canonical, err := canonicalizeJSON(unsigned)
	if err != nil {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	sig, err := base64.RawURLEncoding.DecodeString(signature)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	result := s.verify(func(pub []byte) bool {
		return ed25519.Verify(ed25519.PublicKey(pub), []byte("tulpa/sign\n"+canonical), sig)
	}, artifactMs)
	if !result.Verified {
		return AttestationReasonSignature, MultiKeyResult{}
	}

	issuedAt, _ := obj["issuedAt"].(string)
	expiresAt, _ := obj["expiresAt"].(string)
	start, okStart := ParseInkTimestampMs(issuedAt)
	end, okEnd := ParseInkTimestampMs(expiresAt)
	if !okStart || !okEnd {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	nowMs, okNow := ParseInkTimestampMs(now)
	if !okNow {
		return AttestationReasonSchema, MultiKeyResult{}
	}
	if nowMs < start {
		return AttestationReasonNotYetValid, MultiKeyResult{}
	}
	if nowMs >= end {
		return AttestationReasonExpired, MultiKeyResult{}
	}
	return AttestationReasonNone, result
}

// validateClaimTypeSet enforces the claim-type set shape shared by the
// evidence policy and the structured refusal: an array of 1 to 32 distinct
// strings under the claim-type grammar and bounds.
func validateClaimTypeSet(v interface{}) bool {
	arr, ok := v.([]interface{})
	if !ok || len(arr) < 1 || len(arr) > 32 {
		return false
	}
	seen := make(map[string]bool, len(arr))
	for _, e := range arr {
		s, ok := e.(string)
		if !ok || utf16Len(s) < claimTypeMin || utf16Len(s) > claimTypeMax || !claimTypeRe.MatchString(s) {
			return false
		}
		if seen[s] {
			return false
		}
		seen[s] = true
	}
	return true
}

// validateCardAttestations enforces the Agent Card attestations member: an
// array of 1 to 16 entries, each a well-formed attestation by shape and
// signature grammar. Signature verification and the validity window against a
// clock are verify-time decisions, not card validation: a card carrying a
// stale but well-formed attestation stays valid.
func validateCardAttestations(v interface{}) bool {
	arr, ok := v.([]interface{})
	if !ok || len(arr) < 1 || len(arr) > 16 {
		return false
	}
	for _, e := range arr {
		obj, ok := e.(map[string]interface{})
		if !ok {
			return false
		}
		sig, ok := validateAttestation(obj)
		if !ok || !signatureRe.MatchString(sig) {
			return false
		}
	}
	return true
}

// validateEvidencePolicy enforces the Agent Card evidencePolicy member: an
// object whose optional required and preferred members are claim-type sets.
// Unknown members inside it are ignored for interpretation, matching the
// reference schema's passthrough.
func validateEvidencePolicy(v interface{}) bool {
	p, ok := v.(map[string]interface{})
	if !ok {
		return false
	}
	for _, k := range []string{"required", "preferred"} {
		if pv, present := p[k]; present {
			if !validateClaimTypeSet(pv) {
				return false
			}
		}
	}
	return true
}

// ValidateEvidenceRefusal validates the structured policy:evidence_required
// refusal body of specs/ink-attestation.md, mirroring the reference
// parseEvidenceRefusal: the standard endpoint error members are pinned, the
// claim-type set is bounded and distinct, and unknown members are tolerated
// for forward compatibility.
func ValidateEvidenceRefusal(m map[string]interface{}) bool {
	if m == nil {
		return false
	}
	if s, _ := m["protocol"].(string); s != "ink/0.1" {
		return false
	}
	if b, ok := m["error"].(bool); !ok || !b {
		return false
	}
	if s, _ := m["code"].(string); s != "policy:evidence_required" {
		return false
	}
	v, present := m["requiredClaimTypes"]
	if !present || !validateClaimTypeSet(v) {
		return false
	}
	if mv, present := m["message"]; present {
		s, ok := mv.(string)
		if !ok || utf16Len(s) > 500 {
			return false
		}
	}
	return true
}

// validateAttestation enforces the strict shape: exactly the ten members, the
// single vendor-neutral wire spelling, the claim-type and attestation-id
// grammars, the id bounds and a strictly positive validity window. It returns
// the signature string on success.
func validateAttestation(obj map[string]interface{}) (string, bool) {
	for k := range obj {
		if !attestationTopLevelKeys[k] {
			return "", false
		}
	}
	for _, k := range attestationRequiredKeys {
		if _, ok := obj[k]; !ok {
			return "", false
		}
	}
	if s, _ := obj["protocol"].(string); s != "ink/0.1" {
		return "", false
	}
	if s, _ := obj["type"].(string); s != "network.ink.attestation" {
		return "", false
	}
	issuer, ok := obj["issuer"].(string)
	if !ok || issuer == "" || utf16Len(issuer) > attestationIDMaxs {
		return "", false
	}
	subject, ok := obj["subject"].(string)
	if !ok || subject == "" || utf16Len(subject) > attestationIDMaxs {
		return "", false
	}
	claimType, ok := obj["claimType"].(string)
	if !ok || utf16Len(claimType) < claimTypeMin || utf16Len(claimType) > claimTypeMax || !claimTypeRe.MatchString(claimType) {
		return "", false
	}
	if _, ok := obj["claim"].(map[string]interface{}); !ok {
		return "", false
	}
	attestationID, ok := obj["attestationId"].(string)
	if !ok || utf16Len(attestationID) < attestationIDMin || utf16Len(attestationID) > attestationIDMax || !attestationIDRe.MatchString(attestationID) {
		return "", false
	}
	issuedAt, ok := obj["issuedAt"].(string)
	if !ok {
		return "", false
	}
	expiresAt, ok := obj["expiresAt"].(string)
	if !ok {
		return "", false
	}
	start, okStart := ParseInkTimestampMs(issuedAt)
	end, okEnd := ParseInkTimestampMs(expiresAt)
	if !okStart || !okEnd || end <= start {
		return "", false
	}
	signature, ok := obj["signature"].(string)
	if !ok {
		return "", false
	}
	return signature, true
}

File: src/models/attestation.ts
import { z } from "zod";
import { isInkTimestamp, parseInkTimestampMs } from "../crypto/timestamp.js";
import { isWithinBounds, signMessage, verifyMessage } from "../crypto/sign.js";
import { hasUnpairedSurrogate } from "../crypto/surrogate.js";
import { parseSignedBodyBytes } from "../crypto/parse-signed-body.js";
import { verifyDetachedSignatureWithKeys, type MultiKeyVerifyResult } from "../crypto/multi-key-verify.js";
import type { CandidateKey } from "./key-entry.js";

// A signed claim by one principal about another, the evidence primitive of
// specs/ink-attestation.md. It is not a capability: it binds no audience, no
// presenter and no scope, and base verification passes no judgment on the
// issuer or the claim. Whether a verified attestation means anything is the
// receiver's policy decision, made after verification.

const ID_MAX = 512;
const ATTESTATION_ID_MIN = 16;
const ATTESTATION_ID_MAX = 256;
const CLAIM_TYPE_MIN = 3;
const CLAIM_TYPE_MAX = 128;

// Grammar per the spec: a lowercase reverse-DNS-style dotted name for claim
// types, the shared nonce grammar for attestation ids.
const CLAIM_TYPE_RE = /^[a-z0-9]+(\.[a-z0-9_]+)+$/;
const ATTESTATION_ID_RE = /^[A-Za-z0-9_-]+$/;

/**
 * Byte-length ceiling on a raw attestation body, enforced before decoding, in
 * the same position and for the same reason as the grant cap: JSON permits
 * unbounded whitespace between tokens and whitespace vanishes at
 * canonicalization, so a padded body still carries a verifying signature and
 * must be refused on size alone.
 */
export const MAX_ATTESTATION_BODY_BYTES = 65536;

// The window must be strictly positive. Unlike a grant there is no maximum
// lifetime: a claim about a subject is not a capability, and a receiver
// discounts long windows as policy rather than the schema refusing them.
function isWindowPositive(a: { issuedAt: string; expiresAt: string }): boolean {
  const start = parseInkTimestampMs(a.issuedAt);
  const end = parseInkTimestampMs(a.expiresAt);
  if (start === null || end === null) return false;
  return end > start;
}

export const AttestationSchema = z
  .object({
    protocol: z.literal("ink/0.1"),
    // Single spelling by design: the object postdates the namespace migration,
    // so no legacy network.tulpa.* form exists to accept.
    type: z.literal("network.ink.attestation"),
    issuer: z.string().min(1).max(ID_MAX),
    subject: z.string().min(1).max(ID_MAX),
    claimType: z.string().min(CLAIM_TYPE_MIN).max(CLAIM_TYPE_MAX).regex(CLAIM_TYPE_RE),
    // Opaque to the base verifier: bounds-checked and canonicalized, never
    // interpreted. Meaning belongs to the claim type.
    claim: z.record(z.string(), z.unknown()),
    attestationId: z
      .string()
      .min(ATTESTATION_ID_MIN)
      .max(ATTESTATION_ID_MAX)
      .regex(ATTESTATION_ID_RE),
    issuedAt: z.string().refine(isInkTimestamp, { message: "invalid INK timestamp" }),
    expiresAt: z.string().refine(isInkTimestamp, { message: "invalid INK timestamp" }),
    signature: z.string().regex(/^[A-Za-z0-9_-]{86}$/),
  })
  .strict()
  .refine(isWindowPositive, { message: "expiresAt must be strictly after issuedAt" });

const UnsignedAttestationSchema = z
  .object({
    protocol: z.literal("ink/0.1"),
    type: z.literal("network.ink.attestation"),
    issuer: z.string().min(1).max(ID_MAX),
    subject: z.string().min(1).max(ID_MAX),
    claimType: z.string().min(CLAIM_TYPE_MIN).max(CLAIM_TYPE_MAX).regex(CLAIM_TYPE_RE),
    claim: z.record(z.string(), z.unknown()),
    attestationId: z
      .string()
      .min(ATTESTATION_ID_MIN)
      .max(ATTESTATION_ID_MAX)
      .regex(ATTESTATION_ID_RE),
    issuedAt: z.string().refine(isInkTimestamp, { message: "invalid INK timestamp" }),
    expiresAt: z.string().refine(isInkTimestamp, { message: "invalid INK timestamp" }),
  })
  .strict()
  .refine(isWindowPositive, { message: "expiresAt must be strictly after issuedAt" });

export type Attestation = z.infer<typeof AttestationSchema>;

export const ClaimTypeSchema = z
  .string()
  .min(CLAIM_TYPE_MIN)
  .max(CLAIM_TYPE_MAX)
  .regex(CLAIM_TYPE_RE);

// A claim-type array is a set: distinct entries, satisfied by one credited
// attestation of the type, never by a count.
const ClaimTypeSetSchema = z
  .array(ClaimTypeSchema)
  .min(1)
  .max(32)
  .refine((s) => new Set(s).size === s.length, { message: "claim types must be distinct" });

// The receiver's advance evidence statement on its card. Unknown members pass
// through so the card proof covers them; they carry no meaning here.
export const EvidencePolicySchema = z
  .object({
    required: ClaimTypeSetSchema.optional(),
    preferred: ClaimTypeSetSchema.optional(),
  })
  .passthrough();

export type EvidencePolicy = z.infer<typeof EvidencePolicySchema>;

// The structured refusal a receiver returns when required evidence is missing:
// the standard endpoint error body carrying the policy:evidence_required code
// and the conjunctive residual set of missing claim types. A sender parses it
// from an arbitrary receiver, so unknown members pass through.
export const EvidenceRefusalSchema = z
  .object({
    protocol: z.literal("ink/0.1"),
    error: z.literal(true),
    code: z.literal("policy:evidence_required"),
    requiredClaimTypes: ClaimTypeSetSchema,
    message: z.string().max(500).optional(),
  })
  .passthrough();

export type EvidenceRefusal = z.infer<typeof EvidenceRefusalSchema>;

export type EvidenceRefusalParseResult =
  | { ok: true; refusal: EvidenceRefusal }
  | { ok: false };

/** Parse a candidate evidence refusal body. Never throws. */
export function parseEvidenceRefusal(value: unknown): EvidenceRefusalParseResult {
  const parsed = EvidenceRefusalSchema.safeParse(value);
  return parsed.success ? { ok: true, refusal: parsed.data } : { ok: false };
}

export interface AttestationInput {
  issuer: string;
  subject: string;
  claimType: string;
  claim: Record<string, unknown>;
  attestationId: string;
  issuedAt: string;
  expiresAt: string;
}

export type AttestationVerifyResult =
  | { ok: true; attestation: Attestation }
  | { ok: false; reason: "schema" | "signature" | "not_yet_valid" | "expired" };

export interface AttestationVerifyContext {
  /** The verifier clock, a strict INK timestamp. */
  now: string;
}

/**
 * Build a signed attestation. The unsigned object is validated before signing,
 * so an out-of-grammar claim type or an inverted window is refused at build
 * time rather than producing a signature over an out-of-profile document.
 */
export async function buildAttestation(
  input: AttestationInput,
  issuerPrivateKey: Uint8Array,
): Promise<Attestation> {
  const unsigned = {
    protocol: "ink/0.1" as const,
    type: "network.ink.attestation" as const,
    issuer: input.issuer,
    subject: input.subject,
    claimType: input.claimType,
    claim: input.claim,
    attestationId: input.attestationId,
    issuedAt: input.issuedAt,
    expiresAt: input.expiresAt,
  };
  UnsignedAttestationSchema.parse(unsigned);
  if (!isWithinBounds(unsigned)) {
    throw new Error("attestation exceeds the signed-body structural bounds");
  }
  const signature = await signMessage(unsigned, issuerPrivateKey);
  return { ...unsigned, signature };
}

export type AttestationVerifyWithKeysResult = AttestationVerifyResult & Partial<MultiKeyVerifyResult>;

/**
 * Shared raw-bytes gate, schema parse, surrogate scan, signature step
 * (delegated to `verifySignature`), and validity-window check for an
 * attestation. `beforeSignature`, when given, runs after the surrogate scan
 * and before the signature step; returning a non-null reason fails closed
 * without trying the signature. `verifyAttestation` and
 * `verifyAttestationWithKeys` differ only in how the signature is checked
 * (and, for the WithKeys path, the extra `beforeSignature` artifact-clock
 * gate), so both delegate here.
 */
async function verifyAttestationCore(
  raw: Uint8Array,
  context: AttestationVerifyContext,
  verifySignature: (attestation: Attestation) => Promise<MultiKeyVerifyResult>,
  beforeSignature?: (attestation: Attestation) => "schema" | null,
): Promise<AttestationVerifyWithKeysResult> {
  try {
    if (!ArrayBuffer.isView(raw) || !(raw instanceof Uint8Array)) {
      return { ok: false, reason: "schema" };
    }
    if (raw.length > MAX_ATTESTATION_BODY_BYTES) {
      return { ok: false, reason: "schema" };
    }
    let value: unknown;
    try {
      value = parseSignedBodyBytes(raw);
    } catch {
      return { ok: false, reason: "schema" };
    }
    if (!isWithinBounds(value)) {
      return { ok: false, reason: "schema" };
    }
    const parsed = AttestationSchema.safeParse(value);
    if (!parsed.success) {
      return { ok: false, reason: "schema" };
    }
    const attestation = parsed.data;
    if (hasUnpairedSurrogate(attestation)) {
      return { ok: false, reason: "schema" };
    }
    if (beforeSignature) {
      const problem = beforeSignature(attestation);
      if (problem !== null) {
        return { ok: false, reason: problem };
      }
    }
    const result = await verifySignature(attestation);
    if (!result.verified) {
      return { ok: false, reason: "signature" };
    }
    const start = parseInkTimestampMs(attestation.issuedAt);
    const end = parseInkTimestampMs(attestation.expiresAt);
    if (start === null || end === null) {
      return { ok: false, reason: "schema" };
    }
    const nowMs = parseInkTimestampMs(context.now);
    if (nowMs === null) {
      return { ok: false, reason: "schema" };
    }
    if (nowMs < start) {
      return { ok: false, reason: "not_yet_valid" };
    }
    if (nowMs >= end) {
      return { ok: false, reason: "expired" };
    }
    return result.keyId !== undefined
      ? {
          ok: true,
          attestation,
          keyId: result.keyId,
          keyStatus: result.keyStatus,
          usedRetiredKey: result.usedRetiredKey,
        }
      : { ok: true, attestation };
  } catch {
    return { ok: false, reason: "schema" };
  }
}

/**
 * Verify an attestation from its raw bytes against the resolved issuer key.
 * The input is the raw body, never a parsed value: the raw-body gate of
 * ink-signed-string-safety.md runs on bytes that no longer exist after
 * parsing. Check order per the spec, first failure wins: raw gate and parse,
 * structural bounds and grammar, signature, validity window (lower bound
 * inclusive, upper bound exclusive). Deliberately absent: audience, replay
 * and any judgment about the issuer or the claim. Never throws.
 */
export async function verifyAttestation(
  raw: Uint8Array,
  issuerPublicKey: Uint8Array,
  context: AttestationVerifyContext,
): Promise<AttestationVerifyResult> {
  return verifyAttestationCore(raw, context, async (attestation) => ({
    verified: await verifyMessage(attestation, issuerPublicKey),
  }));
}

/**
 * Verify an attestation from its raw bytes against a rotation-aware
 * candidate issuer key set. Security considerations §"Issuer key rotation"
 * (ink-attestation.md): an attestation verifies under the same rotation
 * rules as any other signed body, a retired issuer key still verifies an
 * attestation issued inside that key's validity window, and a revoked
 * issuer key never verifies, even for an attestation whose `issuedAt`
 * predates the revocation.
 *
 * The artifact clock is `issuedAt`, the moment the issuer signed the
 * claim, i.e. the same field the validity-window check below uses as its
 * lower bound, not `context.now` (the verifier's clock, used only to
 * judge freshness against `expiresAt`). Mirrors `verifyAttestation` byte
 * for byte otherwise; only the signature step is rotation-aware.
 */
export async function verifyAttestationWithKeys(
  raw: Uint8Array,
  keys: CandidateKey[],
  context: AttestationVerifyContext,
  opts?: { hintKeyId?: string },
): Promise<AttestationVerifyWithKeysResult> {
  return verifyAttestationCore(
    raw,
    context,
    (attestation) => {
      // Non-null: `beforeSignature` below already rejected a null parse.
      const artifactMs = parseInkTimestampMs(attestation.issuedAt)!;
      return verifyDetachedSignatureWithKeys(
        (publicKey) => verifyMessage(attestation, publicKey),
        keys,
        artifactMs,
        opts?.hintKeyId,
      );
    },
    (attestation) => (parseInkTimestampMs(attestation.issuedAt) === null ? "schema" : null),
  );
}
